### PR TITLE
Turning off teleport beacon actually disables teleporting to it and adds unit tests for teleporter

### PR DESCRIPTION
**Pull request by Das15**
## About The Pull Request
Teleport beacon can be turned on and off. Turning off only removes it from the list of selectable locations on the teleporter, you can still teleport to it. In this PR <s>I add weak_ref removal on turning off the beacon (of which I'm not certain if that's a good solution, but seems to work)</s> I use signals to handle turning off the teleportation.
## Why It's Good For The Game
You can turn off the beacon to stop someone teleporting to your location and it doesn't mislead players into thinking disabled teleport beacon cannot be teleported to anymore.
## Changelog
:cl:
fix: You can no longer teleport to disabled beacon if the teleporter was previously locked-on to it.
/:cl:

### CODE REVIEW

File: code/game/objects/items/devices/beacon.dm [hunk 1]
```diff
@@ -20,14 +20,18 @@
 	GLOB.teleportbeacons -= src
 	return ..()
 
+/obj/item/beacon/proc/turn_off()
+	icon_state = "beacon-off"
+	GLOB.teleportbeacons -= src
+	src.weak_reference = null
```
Comment (MrMelbert) [inline code/game/objects/items/devices/beacon.dm:26]: I don't think anything should be `null`ing its own weakref, this should be fixed in some other way

One idea: Send a signal in `/obj/machinery/teleport/hub/proc/teleport` to `target`, something like `COMSIG_MOVABLE_BEING_TELEPORTED_TO`, then have beacons hook the signal and return a blocking flag if disabled

Another idea, have beacons send a signal when turned off `COMSIG_BEACON_DISABLED` that teleport hubs can hook onto and clear the target if caught

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/beacon.dm (hunk 1 above)
Comment (Das15) [reply code/game/objects/items/devices/beacon.dm:26]: Yup, seems a lot better idea than just outright removing weak_ref. I'm going to implement it.

**Comment by carlarctg** [comment]
give this man the unit test label

File: code/game/objects/items/devices/beacon.dm [hunk 2]
```diff
@@ -20,14 +20,18 @@
...
+
 /obj/item/beacon/attack_self(mob/user)
 	enabled = !enabled
 	if (enabled)
 		icon_state = "beacon"
 		GLOB.teleportbeacons += src
 	else
-		icon_state = "beacon-off"
-		GLOB.teleportbeacons -= src
+		src.turn_off()
```
Comment (san7890) [inline code/game/objects/items/devices/beacon.dm:34]:
```suggestion
		turn_off()
```

feels rather redundant

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/beacon.dm (hunk 2 above)
Comment (Das15) [reply code/game/objects/items/devices/beacon.dm:34]: Okay, removing that.

File: code/game/machinery/computer/teleporter.dm [hunk 3]
```diff
@@ -110,6 +122,11 @@
 		return
 	SEND_SIGNAL(src, COMSIG_TELEPORTER_NEW_TARGET, new_target)
 	target_ref = new_target_ref
+	if (istype(new_target, /obj/item/beacon))
```
Comment (Jacquerel) [inline code/game/machinery/computer/teleporter.dm:125]:
```suggestion
	if (!istype(new_target, /obj/item/beacon))
		return
```

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Tiny suggestion

File: code/game/machinery/computer/teleporter.dm (hunk 3 above)
Comment (Das15) [reply code/game/machinery/computer/teleporter.dm]: Seems like a good idea.

File: code/game/machinery/computer/teleporter.dm [hunk 4]
```diff
@@ -111,6 +124,13 @@
 	SEND_SIGNAL(src, COMSIG_TELEPORTER_NEW_TARGET, new_target)
 	target_ref = new_target_ref
 
+	if (!istype(new_target, /obj/item/beacon))
+		return
+	if (has_registered_signal)
+		UnregisterSignal(target_ref.resolve(), COMSIG_BEACON_DISABLED)
```
Comment (Jacquerel) [inline code/game/machinery/computer/teleporter.dm:130]: Does this work? You're doing it _after_ changing what `target_ref` points at.
Would this not be simpler if you just:
- Check if `target_ref` resolves to anything
- If yes and if it is the same atom as `new_target` then return, we don't need to do anything
- Unregister the signal from `target_ref`
- Reassign `target_ref` to `new_target_ref`
- Register signal on `new_target`

That would probably also get rid of the need for the `has_registered_signal` flag?

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/game/machinery/computer/teleporter.dm (hunk 4 above)
Comment (Das15) [reply code/game/machinery/computer/teleporter.dm:130]: This seems like a good idea. I've tried to implement something along those lines.

Speaking about this working, I've tested again and everything except removing signals works. It only removes beacon signal if beacon was turned off. I wonder why UnregisterSignal doesn't runtime or write to logs when you try to unregister non-registered atom.

**Pull request merged**